File: rust/src/bin/contains_duplicate.rs
use std::collections::{BTreeSet, HashSet};
pub struct Solution;

impl Solution {
    // With HashMap<i32,()> there is incredibly small chance of hash collision so one would need HashMap<i32,i32> that
    // put a as both key and value and then compares values. Which is exactly what Hashset is under
    // the hood
    pub fn hash_simple(nums: &[i32]) -> bool {
        let mut seen_elements = HashSet::new();
        for &a in nums {
            if !seen_elements.insert(a) {
                return true;
            }
        }
        return false;
    }
    pub fn hash_iter(nums: &[i32]) -> bool {
        let mut seen = HashSet::new();
        nums.iter().any(|&a| !seen.insert(a))
    }
    // suboptimal to hashset
    pub fn bst_iter(nums: &[i32]) -> bool {
        let mut seen = BTreeSet::new();
        nums.iter().any(|&a| !seen.insert(a))
    }
    // Sorting way O(N log N) but with O(1) space
    // requires mutable vector to sort it in place
    pub fn sort_simple(mut nums: Vec<i32>) -> bool {
        nums.sort_unstable();
        nums.windows(2).any(|w| w.get(0) == w.get(1))
    }
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple() {
        let num = vec![1, 2, 3, 1];
        let num2 = vec![1, 2, 3, 4];
        assert_eq!(Solution::hash_simple(&num), true);
        assert_eq!(Solution::hash_simple(&num2), false);
    }
    #[test]
    fn test_iter() {
        let num = vec![1, 2, 3, 1];
        let num2 = vec![1, 2, 3, 4];
        assert_eq!(Solution::hash_iter(&num), true);
        assert_eq!(Solution::hash_iter(&num2), false);
    }

    fn test_suite(f: fn(&[i32]) -> bool) {
        let num = vec![1, 2, 3, 1];
        let num2 = vec![1, 2, 3, 4];
        assert_eq!(f(&num), true);
        assert_eq!(f(&num2), false);
    }

    #[test]
    fn test_hash_simple() {
        test_suite(Solution::hash_simple);
    }
    #[test]
    fn test_hash_iter() {
        test_suite(Solution::hash_iter);
    }
    #[test]
    fn test_bst() {
        test_suite(Solution::bst_iter);
    }

    // Macro way
    macro_rules! generate_tests {
        ($($name:ident),* => $version:literal) => {
           $(
            paste::paste! {
                #[test]
                fn [< test_ $name _ $version >](){
                    let num = vec![1, 2, 3, 1];
                    let num2 = vec![1, 2, 3, 4];
                    let f = Solution::$name;
                    assert_eq!(f(&num),true);
                    assert_eq!(f(&num2),false);
                }
            }
           )*
        };
    }

    generate_tests!(hash_simple, hash_iter => 2);
}
pub fn main() {}

File: cpp/contains_duplicate.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
// #include <set>
#include <algorithm>
#include <cassert>

using namespace std;


class Solution {
public:
	static bool hash_simple(vector<int> &nums) {
		unordered_set<int> seen;
		for (int a : nums){
			if (!seen.insert(a).second)
				return true;
		}
		return false;
	}
	static bool hash_iter(vector<int> &nums){
		unordered_set<int> seen;
		return any_of(nums.begin(), nums.end(), [&seen](int a){
		return !seen.insert(a).second;
		});
	}
	static bool sort_simple(vector<int> &nums) {
		sort(nums.begin(),nums.end());
		return adjacent_find(nums.begin(),nums.end()) !=nums.end();
	}
};

void test_suite(bool (*f)(vector<int> &), const string& name){
	vector<int> num1 = {1,2,3,1};
	vector<int> num2 = {1,2,3,4};

	assert(f(num1) == true);
	assert(f(num2) == false);

	cout<<name<<" passed"<<endl;
}


int main() {
	test_suite(Solution::hash_simple, "hash_simple");
	test_suite(Solution::hash_iter, "hash_iter");
	test_suite(Solution::sort_simple, "sort_simple");
	return 0;
}
